src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala

```diff
@@ -0,0 +1,343 @@
+    provider match {
+      case "delta" =>
+        new StagedDeltaTableV2(ident, schema, partitions, properties)
+      case _ =>
+        throw new IllegalStateException("not supported yet")
+    }
+  }
+
+  // Copy of V2SessionCatalog.convertTransforms, which is private.
+  private def convertTransforms(partitions: Seq[Transform]): (Seq[String], Option[BucketSpec]) = {
+    val identityCols = new mutable.ArrayBuffer[String]
+    var bucketSpec = Option.empty[BucketSpec]
+
+    partitions.map {
+      case IdentityTransform(FieldReference(Seq(col))) =>
+        identityCols += col
+
+      case BucketTransform(numBuckets, FieldReference(Seq(col))) =>
+        bucketSpec = Some(BucketSpec(numBuckets, col :: Nil, Nil))
+
+      case transform =>
+        throw new UnsupportedOperationException(
+          s"SessionCatalog does not support partition transform: $transform")
+    }
+
+    (identityCols, bucketSpec)
+  }
+
+  // Copy-pasted from DeltaAnalysis.
+  private def verifyTableAndSolidify(
+      tableDesc: CatalogTable,
+      query: Option[LogicalPlan]): CatalogTable = {
+
+    if (tableDesc.bucketSpec.isDefined) {
+      throw DeltaErrors.operationNotSupportedException("Bucketing", tableDesc.identifier)
+    }
+
+    val schema = query.map { plan =>
+      assert(tableDesc.schema.isEmpty, "Can't specify table schema in CTAS.")
+      plan.schema.asNullable
+    }.getOrElse(tableDesc.schema)
+
+    PartitioningUtils.validatePartitionColumn(
+      schema,
+      tableDesc.partitionColumnNames,
+      caseSensitive = false) // Delta is case insensitive
+
+    val validatedConfigurations = DeltaConfigs.validateConfigurations(tableDesc.properties)
+
+    val db = tableDesc.identifier.database.getOrElse(catalog.getCurrentDatabase)
+    val tableIdentWithDB = tableDesc.identifier.copy(database = Some(db))
+    tableDesc.copy(
+      identifier = tableIdentWithDB,
+      schema = schema,
+      properties = validatedConfigurations)
+  }
+
+  // Copy-pasted from DeltaAnalysis.
+  private def getExistingTableIfExists(table: CatalogTable): Option[CatalogTable] = {
+    val tableExists = catalog.tableExists(table.identifier)
+    if (tableExists) {
+      val oldTable = catalog.getTableMetadata(table.identifier)
+      if (oldTable.tableType == CatalogTableType.VIEW) {
+        throw new AnalysisException(
+          s"${table.identifier} is a view. You may not write data into a view.")
+      }
+      // TODO(burak): Maybe drop old table if mode is overwrite?
+      if (!DeltaSourceUtils.isDeltaTable(oldTable.provider)) {
+        throw new AnalysisException(s"${table.identifier} is not a Delta table. Please drop this " +
+          "table first if you would like to create it with Databricks Delta.")
+      }
+      Some(oldTable)
+    } else {
+      None
+    }
+  }
+
+  // We keep this as an internal class because it needs to call into some catalog internals for the
+  // eventual table creation.
+  private class StagedDeltaTableV2(
```

Comment:
I would change the logic in this class a little bit if possible, but we can do that in a follow up.